atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketWriter.java

```diff
@@ -0,0 +1,110 @@
+import com.palantir.atlasdb.sweep.queue.ShardAndStrategy;
+import com.palantir.atlasdb.table.description.SweeperStrategy;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.List;
+import java.util.Optional;
+
+final class DefaultBucketWriter implements BucketWriter {
+    private static final SafeLogger log = SafeLoggerFactory.get(DefaultBucketWriter.class);
+    private final SweepBucketsTable sweepBucketsTable;
+    private final List<SweeperStrategy> sweeperStrategies;
+    private final int numberOfShards;
+
+    // The list ordering needs to be consistent, even when adding new sweeper strategies. New strategies must be added
+    // to the end.
+    private DefaultBucketWriter(
+            SweepBucketsTable sweepBucketsTable, List<SweeperStrategy> sweeperStrategies, int numberOfShards) {
+        this.sweepBucketsTable = sweepBucketsTable;
+        this.sweeperStrategies = sweeperStrategies;
+        this.numberOfShards = numberOfShards;
+    }
+
+    // A static number of shards, since it will not be configurable anymore.
+    static BucketWriter create(
+            SweepBucketsTable sweepBucketsTable, List<SweeperStrategy> sweeperStrategies, int numberOfShards) {
+        return new DefaultBucketWriter(sweepBucketsTable, sweeperStrategies, numberOfShards);
+    }
+
+    @Override
+    public WriteState writeToAllBuckets(
+            long bucketIdentifier, Optional<TimestampRange> oldTimestampRange, TimestampRange newTimestampRange) {
+
+        log.info(
+                "Assigning bucket ID {} timestamp range {} from old range {}",
+                SafeArg.of("bucketIdentifier", bucketIdentifier),
+                SafeArg.of("newTimestampRange", newTimestampRange),
+                SafeArg.of("oldTimestampRange", oldTimestampRange));
+        boolean failedAtStart = false;
+        // A consistent ordering (by shard number and strategy), so that retries and concurrent threads are CASing
+        // in the same order, making it easier to determine if this is potentially a retry (there was a failure at the
+        // start, implying there was a previous write), or contention (we were successful at the beginning so no
+        // previous successful attempts at writing, but now failing implying someone else is updating).
+        for (SweeperStrategy sweeperStrategy : sweeperStrategies) {
+            for (int shard = 0; shard < numberOfShards; shard++) {
+                Bucket bucket = Bucket.of(ShardAndStrategy.of(shard, sweeperStrategy), bucketIdentifier);
+                try {
+                    sweepBucketsTable.putTimestampRangeForBucket(bucket, oldTimestampRange, newTimestampRange);
+                    // reset the flag after a successful attempt, so we can differentiate between a potential abort
+                    // and contention.
+                    failedAtStart = false;
```

Comment:
yeah I think this makes sense, though maybe let's rename this to indicate that it's part of processing a failing _prefix_ of the buckets?